pkg/controllers/management/k3supgrade/register.go

```diff
@@ -0,0 +1,209 @@
+	"fmt"
+	"reflect"
+
+	"github.com/rancher/rancher/pkg/clustermanager"
+	"github.com/rancher/rancher/pkg/systemaccount"
+	"github.com/rancher/rancher/pkg/wrangler"
+	wranglerv3 "github.com/rancher/rancher/pkg/wrangler/generated/controllers/management.cattle.io/v3"
+	planv1 "github.com/rancher/system-upgrade-controller/pkg/apis/upgrade.cattle.io/v1"
+	planClientset "github.com/rancher/system-upgrade-controller/pkg/generated/clientset/versioned/typed/upgrade.cattle.io/v1"
+	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
+	projectv3 "github.com/rancher/types/apis/project.cattle.io/v3"
+	"github.com/rancher/types/config"
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type handler struct {
+	systemUpgradeNamespace string
+	clusterCache           wranglerv3.ClusterCache
+	clusterClient          wranglerv3.ClusterClient
+	apps                   projectv3.AppInterface
+	appLister              projectv3.AppLister
+	templateLister         v3.CatalogTemplateLister
+	systemAccountManager   *systemaccount.Manager
+	manager                *clustermanager.Manager
+}
+
+const (
+	systemUpgradeNS        = "cattle-system"
+	rancherManagedPlan     = "rancher-managed"
+	upgradeDisableLabelKey = "plan.upgrade.cattle.io/disable"
+	k3sUpgraderCatalogName = "system-library-rancher-k3s-upgrader"
+)
+
+func Register(ctx context.Context, wContext *wrangler.Context, mgmtCtx *config.ManagementContext, manager *clustermanager.Manager) {
+	h := &handler{
+		systemUpgradeNamespace: systemUpgradeNS,
+		clusterCache:           wContext.Mgmt.Cluster().Cache(),
+		clusterClient:          wContext.Mgmt.Cluster(),
+		apps:                   mgmtCtx.Project.Apps(metav1.NamespaceAll),
+		appLister:              mgmtCtx.Project.Apps("").Controller().Lister(),
+		templateLister:         mgmtCtx.Management.CatalogTemplates("").Controller().Lister(),
+		systemAccountManager:   systemaccount.NewManager(mgmtCtx),
+		manager:                manager,
+	}
+
+	wContext.Mgmt.Cluster().OnChange(ctx, "k3s-upgrade-controller", h.onClusterChange)
+}
+
+// deployPlans creates a master and worker plan in the downstream cluster to instrument
+// the system-upgrade-controller in the downstream cluster
+func (h *handler) deployPlans(cluster *v3.Cluster) error {
+
+	// access downstream cluster
+	clusterCtx, err := h.manager.UserContext(cluster.Name)
+	if err != nil {
+		return err
+
+	}
+
+	// create a client for GETing Plans in the downstream cluster
+	// TODO: We shouldn't create one every time
+	planConfig, err := planClientset.NewForConfig(&clusterCtx.RESTConfig)
+	if err != nil {
+		return err
+	}
+	planClient := planConfig.Plans(metav1.NamespaceAll)
+
+	planList, err := planClient.List(metav1.ListOptions{})
+	if err != nil {
+		// may need to handle this error, if there is no Plan CRD what should we do?
+		if errors.IsNotFound(err) {
+			// no plan CRD exists
+			logrus.Warnf("plan CRD does not exist: %s", err)
```

Comment:
nit: maybe just make this the err that is returned in this case and say, the crd does not exist _yet_

Reply:
I don't think it matters what the error here is actually. We should just requeue on any error.

Reply:
In other controllers we may not return the 'NotFound' error to the framework but here we know that the system-upgrade-controller should be installed